sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,188 @@
+	pq.Lock()
+	defer pq.Unlock()
+
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i == nil || i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.Lock()
+	defer pq.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(&pq.data, clone)
+	return nil
+}
+
+// PopByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
+	pq.Lock()
+	defer pq.Unlock()
+
+	item, ok := pq.dataMap[key]
+	if !ok {
+		return nil, nil
+	}
+
+	// remove the item the heap and delete it from the dataMap
+	itemRaw := heap.Remove(&pq.data, item.index)
+	delete(pq.dataMap, key)
+
+	if i, ok := itemRaw.(*Item); ok {
+		return i, nil
+	}
+
+	return nil, nil
+}
+
+// Len returns the number of items in the queue data structure. Do not use this
+// method directly on the queue, use PriorityQueue.Len() instead.
+func (q queue) Len() int { return len(q) }
+
+// Less returns whether the Item with index i should sort before the Item with
+// index j in the queue. This method is used by the queue to determine priority
+// internally; the Item with the lower value wins. (priority zero is higher
+// priority than 1). The priority of Items with equal values is undetermined.
+func (q queue) Less(i, j int) bool {
+	return q[i].Priority < q[j].Priority
+}
+
+// Swap swaps things in-place; part of sort.Interface
+func (q queue) Swap(i, j int) {
+	q[i], q[j] = q[j], q[i]
+	q[i].index = i
+	q[j].index = j
+}
+
+// Push is used by heap.Interface to push items onto the heap. This method is
+// invoked by container/heap, and should not be used directly.
+// See: https://golang.org/pkg/container/heap/#Interface
+func (q *queue) Push(x interface{}) {
+	n := len(*q)
+	item := x.(*Item)
+	item.index = n
+	*q = append(*q, item)
+}
+
+// Pop is used by heap.Interface to pop items off of the heap. This method is
+// invoked by container/heap, and should not be used directly.
+// See: https://golang.org/pkg/container/heap/#Interface
+func (q *queue) Pop() interface{} {
+	old := *q
+	n := len(old)
+	item := old[n-1]
```

Comment:
Looks like they've modified this example recently to avoid a memory leak:  https://github.com/golang/go/commit/7a9968f96fdfa0979e7f086d37592c5df9b9a27c

Reply:
Nice find, thanks! Updated in 47b4a4d77